src/riak_cs_quota.erl

```diff
@@ -0,0 +1,130 @@
+%% under the License.
+%%
+%% ---------------------------------------------------------------------
+
+%% @doc Behaviour for Quota / QoS hook point, to save CS systems by
+%% limiting loads. The idea of Quota and QoS system is, to limit
+%% things by pluggable modules. This is because requirements vastly
+%% varies and it is too hard to cover all of them with single
+%% implementation.
+%% This is experimental feature. These interfaces and behaviousr
+%% may or may not change until general availability.
+%%
+%% To use 'inherited' your own module, it should be configured in
+%% application environment (set it in advanced.config). The item name
+%% is `quota_modules'. Example:
+%%
+%% {riak_cs, [
+%%   {quota_modules, [riak_cs_simple_quota, riak_cs_bw_limitter]},
+%%   ...
+%% ]},
+%%
+%% The order of module names is also the order of evaluation. Modules
+%% are evaluated every time 1) incoming requests passed every other
+%% authentication, authorization works (including ACL, Policy), and 2)
+%% all incoming requests successfully finished even if those requests
+%% resulted in 50x or whatever.
+%%
+%% 1) is a callback called `allow/3' and 2) is `update/2` callback.
+%% These callbacks are called in each process context, 1) is at
+%% mochiweb acceptor and 2) is webmachine log handler. Another
+%% callback `start_link/0' is to let supervisor know the quota module
+%% needs a state management server or not. These processes can't be
+%% added nor removed at runtime, please just control such
+%% enabled/disabled status inside plugin modules.
+%%
+%% Return value of `allow/3' is evaluated at `error_response/3', where
+%% client return behaviour can be controlled here, such as http status
+%% codes, headers and body.
+%%
+%% Known limitations:
+%% - currently this is only for objects.
+%% - separate owner and accessor case: bandwidth and access should be
+%%   for accessor, whild owner is for usage
+%% - no integration tests, please report bugs if you find any!
+%% - No cool configuration systems
+%% - No cool operation console: use `riak-cs attach' and send msg to processes
+%% - Multibag
+
+-module(riak_cs_quota).
+
+-export([invoke_all_callbacks/3,
+         update_all_states/2,
+         handle_error/4]).
+
+-include("riak_cs.hrl").
+-include_lib("webmachine/include/webmachine_logger.hrl").
+-include_lib("webmachine/include/wm_reqdata.hrl").
+
+%% Callbacks as a behaviour
+
+%% @doc initalize a global state of a quota instance. This state is
+%% preserved while this quota process is alive. As no system
+%% configurations provided, module programmers should set their own
+%% configuration system. So aren't command-line console.
+-callback start_link() ->
+    {ok, pid()} | ignore | {error, Reason :: term()}.
+
+%% @doc note that this callback is not called at single process, but
+%% called at each mochiweb acceptor context.
+-callback allow(
+            Owner :: rcs_user(), %% Owner of the target object
```

Comment:
Also there is possibility for `binary()` for non-owner access but allowed by ACL/Policy .